WordPress/Helpers/WPDBTrait.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/**
+ * WordPress Coding Standard.
+ *
+ * @package WPCS\WordPressCodingStandards
+ * @link    https://github.com/WordPress/WordPress-Coding-Standards
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WordPressCS\WordPress\Helpers;
+
+use PHP_CodeSniffer\Files\File;
+
+/**
+ * Helper utilities for sniffs which examine WPDB method calls.
+ *
+ * {@internal The method in this trait was previously contained in the
+ * `WordPressCS\WordPress\Sniff` class and has been moved here.}
+ *
+ * @package WPCS\WordPressCodingStandards
+ * @since   3.0.0
+ */
+trait WPDBTrait {
+
+	/**
+	 * Checks whether this is a call to a $wpdb method that we want to sniff.
+	 *
+	 * If available in the class using this trait, the $methodPtr, $i and $end properties
+	 * are automatically set to correspond to the start and end of the method call.
+	 * The $i property is also set if this is not a method call but rather the
+	 * use of a $wpdb property.
+	 *
+	 * @since 0.8.0
+	 * @since 0.9.0  The return value is now always boolean. The $end and $i member
+	 *               vars are automatically updated.
+	 * @since 0.14.0 Moved this method from the `PreparedSQL` sniff to the base WP sniff.
+	 * @since 3.0.0  - Moved from the Sniff class to this dedicated Trait.
+	 *               - The $phpcsFile parameter was added.
+	 *
+	 * {@internal This method should be refactored to not exhibit "magic" behaviour
+	 *            for properties in the sniff class(es) using it.}}
+	 *
+	 * @param \PHP_CodeSniffer\Files\File $phpcsFile      The file being scanned.
+	 * @param int                         $stackPtr       The index of the $wpdb variable.
+	 * @param array                       $target_methods Array of methods. Key(s) should be method name.
+	 *
+	 * @return bool Whether this is a $wpdb method call.
+	 */
+	protected function is_wpdb_method_call( File $phpcsFile, $stackPtr, $target_methods ) {
+
+		$tokens = $phpcsFile->getTokens();
+
+		// Check for wpdb.
+		if ( ( \T_VARIABLE === $tokens[ $stackPtr ]['code'] && '$wpdb' !== $tokens[ $stackPtr ]['content'] )
+			|| ( \T_STRING === $tokens[ $stackPtr ]['code'] && 'wpdb' !== $tokens[ $stackPtr ]['content'] )
```

Comment:
Under what conditions would a T_STRING of `wpdb` be the equivalent of T_VARIABLE of `$wpdb`? Different PHPCS versions? Use of `$this->wpdb`? Something else?

Reply:
Good question, which should probably have been asked 5 years ago when that code was added.

`$this->wpdb` is a typical one, where a plugin stores the global `$wpdb` variable in a property of their own class instead of having a `global` statement in every method, though in that case, the `strtolower()` should probably not be applied.

Another one would be a static method access `wpdb::methodname()`. And yes, I'm aware that at this time, the WP native `wpdb` class doesn't contain any static methods. May have been different in the past, may also have been put in place for plugins having their own `PluginNamespace\wpdb` class.